File: preprocess_imaris.R
library(readr)
library(pbmcapply)
library(parallel)
library(logging)
options(warn=-1)
basicConfig()

gatherdat=function(fil){
    dataset <- read_csv(fil[1],skip = 3,col_types=cols())
    
    for (f in fil[-1]){
        dataset_temp <- read_csv(f,skip = 3,col_types=cols())
        if (any(grepl("Parent",colnames(dataset)))){
            dataset_temp$Parent=dataset_temp$Parent+max(dataset$Parent)
        }
        dataset_temp$ID=max(dataset$ID)+1
        dataset=rbind(dataset,dataset_temp)
    }
    
    return(dataset)
}

args = commandArgs(trailingOnly=TRUE)

l=list.dirs(args[1],recursive=F)
l=l[!grepl(".ipynb_checkpoints",l)]

scale=read.table(paste0(args[1],"/scale"),header=TRUE)
sc=scale$real_distance/scale$imaris_distance


for (d in l){
    loginfo(paste0("Processing ",d))
    
    allfiles=list(list.files(d,recursive=T,full.names=T,pattern="_tracks_Position"),
              list.files(d,recursive=T,full.names=T,pattern="_tracks_Speed"),
              list.files(d,recursive=T,full.names=T,pattern="_Track_Speed_Max"),
              list.files(d,recursive=T,full.names=T,pattern="_Track_Speed_Mean"))

    loginfo(paste0("Gathering and rescaling all tracks and speeds using ",length(allfiles)," cores"))
    
    datasets=mclapply(allfiles,gatherdat,mc.cores=length(allfiles))
    
    alltracks=datasets[[1]]
    alltracks_tocsv=alltracks;alltracks_tocsv[,1:2]=alltracks_tocsv[,1:2]*sc
    allspeeds=datasets[[2]];allspeeds[,1]=allspeeds[,1]*sc
    allmaxspeeds=datasets[[3]];allmaxspeeds[,1]=allmaxspeeds[,1]*sc
    allmeanspeeds=datasets[[4]];allmeanspeeds[,1]=allmeanspeeds[,1]*sc
    
    write.csv(allspeeds,paste0(d,"/all_speeds.csv"))
    write.csv(allmaxspeeds,paste0(d,"/all_maxspeeds.csv"))
    write.csv(allmeanspeeds,paste0(d,"/all_meanspeeds.csv"))
    
    write.csv(alltracks_tocsv,paste0(d,"/all_tracks.csv"))
    
}

File: preprocess_imaris.py
import glob
import pandas as pd
from p_tqdm import p_map
import sys
patterns=["/**/*_tracks_Position*","/**/*_tracks_Speed*",
         "/**/*_Track_Speed_Max*","/**/*_Track_Speed_Mean*"]

def gather_inaris(fls):
    dataset=pd.read_csv(fls[0],skiprows=3)
    for fl in fls[1:]:
        dataset_temp=pd.read_csv(fl,skiprows=3)
        if any(dataset.columns=="Parent"):
            dataset_temp.Parent=dataset_temp.Parent+max(dataset.Parent)
        dataset_temp.ID=max(dataset.ID)+1
        dataset=dataset.append(dataset_temp)
    if any(dataset.columns=="Value"):
        dataset.loc[:,["Value"]]=dataset.loc[:,["Value"]]*sc
    else:
        dataset.loc[:,["Position X","Position Y"]]=dataset.loc[:,["Position X","Position Y"]]*sc
        
    return dataset

def imaris_to_csv(fold):
    filesets=list(map(lambda x: glob.glob(fold+x),patterns))
    datasets=list(map(gather_inaris,filesets))
    data_tracks, data_speeds, data_maxspeeds, data_meanspeeds = datasets[0], datasets[1], datasets[2], datasets[3]
    data_tracks.to_csv(fold+"all_tracks.csv")
    data_speeds.to_csv(fold+"all_speeds.csv")
    data_maxspeeds.to_csv(fold+"all_maxspeeds.csv")
    data_meanspeeds.to_csv(fold+"all_meanspeeds.csv")
    
    
path=sys.argv[1]
scale=pd.read_table(path+"/scale")
sc=(scale["real_distance"]/scale["imaris_distance"])[0]
folds=glob.glob(path+"/*/")

p_map(imaris_to_csv,folds)
